crates/store/README.md

```diff
@@ -0,0 +1,829 @@
+
+#### Transaction execution
+
+The execution of a transaction typically follows these steps:
+
+  1. A new `Transaction` is created.
+
+  2. Operations (`put()` and `delete()`) are added to the transaction.
+
+  3. The transaction is passed to a layer's `apply()` method.
+
+  4. The layer (often a `Temporal` layer) processes the transaction's
+     operations.
+
+  5. If using a `Temporal` layer, the changes are accumulated until `commit()`
+     is called.
+
+  6. Upon commit, all changes are applied atomically to the underlying store.
+
+#### Consistency and isolation
+
+The transaction system helps ensure consistency and isolation:
+
+  - **Consistency**: By grouping related operations together, transactions help
+    maintain invariants and consistency rules in the data.
+
+  - **Isolation**: Changes made within a transaction (especially when using the
+    `Temporal` layer) are not visible to other parts of the system until the
+    transaction is committed.
+
+#### Performance considerations
+
+The transaction system is designed with performance in mind:
+
+  - The use of a `BTreeMap` for storing operations allows for efficient lookups
+    and range queries.
+
+  - The ability to batch multiple operations into a single transaction can
+    significantly reduce the number of interactions with the underlying storage
+    system.
+
+  - The `Temporal` layer's ability to accumulate changes before committing
+    allows for efficient batching of writes.
+
+#### Extensibility
+
+The transaction system is designed to be extensible:
+
+  - New types of operations could be added to the `Operation` enum if needed
+
+  - Custom layers could implement specialised transaction handling logic.
+
+  - The system could be extended to support more complex transaction patterns,
+    such as nested transactions or savepoints.
+
+#### Error handling
+
+The transaction system is designed to work with Rust's `Result` type and the
+`eyre` error handling library. This allows for detailed error reporting and
+proper error propagation throughout the system.
+
+## Iterators
+
+The store library implements a sophisticated iterator system that provides
+efficient and flexible means of traversing and processing stored data. This
+system is crucial for operations that need to work with ranges of data, such as
+scanning, filtering, or aggregating over large datasets.
+
+### Core iterator concepts
+
+The iterator system is built around several key types and traits:
+
+  1. **`Iter`**: This is the primary iterator type in the library. It's generic
+     over two type parameters, `K` and `V`, which represent the key and value
+     types respectively.
+
+  2. **`DBIter`**: This trait defines the core functionality that any
+     database-specific iterator must implement. It includes methods for moving
+     to the next item (`next()`) and reading the current value (`read()`). In
+     this way it is similar to Rust's `Iterator` trait.
```

Comment:
Another thing to note here about why there's a layer of indirection to reading values is to avoid unnecessary IO.

Unsure how well RocksDB upholds those expectations of permitting key iteration without values considering (ref: #475)

> Worth noting that RocksDB has a boat-load of optimizations, one of which is being able to mmap SST files to avoid memory copying (much akin to our efforts with https://github.com/calimero-network/core/issues/307).
>
> In which case, the PinnableSlice<'_> encapsulated into Slice<'_> from RocksDB::get might just reference an index in these mmap-ed pages, and as such, a further optimization might not perform better.

Reply:
This is covered in the performance section 🙂